### PR TITLE
Support for case-insensitive fields.

### LINKED ISSUE #123: Feature request: case-insensitive keys when Unmarshalling ?
So we can load files like these:

``` yaml
hosts:
  toto:
    hostname: 1.2.3.4
    port: 42
```

As well as

``` yaml
Hosts:
  toto:
    HostName: 1.2.3.4
    Port: 42
```

Using

``` go
type File struct {
  Hosts: []Host `yaml:Hosts`,
}

type Host struct {
  HostName: string `yaml:HostName`
  Port: int `yaml:Port`
}
```

### PR DESCRIPTION
**Pull request by jpap**
Fixes #123.

### CODE REVIEW

**Comment by niemeyer** [comment]
The behavior we have always had is simple and effective, and we don't have a strong reason to change it I think. What we may eventually support is the ability to have logic customizing the decoder/encoder with custom field naming logic. There are existing discussions in PRs and issues about that.

**Comment by jpap** [comment]
It's disappointing that you've closed this PR while the Issue is addresses remains open, and without giving a concrete reason other than the existing codebase is _simple and effective_.  You've talked about unreferenced _existing discussions_ without linking to the PRs and issues.

As for a _strong reason_ -- doesn't Issue #123 provide sufficient motivation for merge?  It's an issue that three developers have voiced concerns about, with @frickenate providing the most spirited feedback.

This unfortunately discourages me from contributing back to your project.  I've been working from a private fork in the last nine months since I submitted this PR, and it takes time and effort to contribute patches back.  I appreciate that it also takes effort for project maintenance on your part, but please help us to help you by providing some actionable feedback when closing PRs without merge, or at by addressing your decision in the associated issue before culling the PR.

Despite all of the above, I really do appreciate the work that you've put into the project.  Your package is extraordinarily useful and I thank you for making it available as open source in the first place. <3

**Comment by niemeyer** [comment]
I apologize it came across as dismissive, as it certainly wasn't my intention. I didn't link to the issue because I'd need to search for it myself. If you could do that on your end, it helps the project, as I can dedicate more of my time slot to fixing actual issues.

**Comment by jpap** [comment]
Thanks for your quick response.

I did take a quick look at existing issues before writing my reply above, and I'm not really sure myself which thread(s) you had in mind.  This is precisely why it is important to provide actionable feedback. ;-)

**Comment by niemeyer** [comment]
List from @cizmazia's PR: #500, #512, #424, #123

**Comment by jpap** [comment]
Thanks for the list. I've just gone through each of them; some comments below.

* #500 (open) talks about key case lowering and being able to marshal key names arbitrarily.  This is a good idea, and I'm all for it; but it brings a heavy-weight tool to the simple problem of case-insensitive field names.  Having a `SetFieldNameTranslator` func for the marshaller, or similar, would likely require the package user to write a lot of code to filter the subset of "case insensitive" fields of interest vs. a simple tag on the struct.  The latter approach of this PR achieves that in a very simple and clean way for the package user.  You could say this PR is a simple solution to a common problem; #500 is for when you want to implement a more complex and custom field naming scheme.
* #512 (open, I'm the author) talks about applying flags across all fields by default.  (I've also submitted a PR for that one, too.)  That's orthogonal to this PR.  I'm using both PRs in a project right now, which is why I've contributed them back here. :)
* #424 (closed) doesn't seem relevant here: it talks about the **tag name** ("json" vs. "yaml"), and not about how the field names are interpreted.
* #123 (open) is the issue that this PR addresses.

The tl;dr of the above: this PR and issue #123 is still relevant. <3

**Comment by niemeyer** [comment]
There's a common topic across these: people want to do custom things to field names and tags. We'll solve that issue in a way that will allow doing things in a custom way, without the library itself having to support combinations of all desired cases internally, and without anyone's preference affecting projects that happen to build together different packages that internally handle their own files using go-yaml.

**Comment by andreimatei** [comment]
I'll kindly pile my own need: I would like camel-cased yaml field names to be unmarshaled into pascal-cased go struct fields (i.e. the difference is in the capitalization of the first letter only. For reference, the stdlib json unmarshaler works how I want it.
E.g.

```go
type Foo struct {
	CollectExpr string
}

yaml := `
collectExpr: xxx
`
```

FWIW, in my case I do not control the generation of the `structs`, so I cannot add tags.
